server/backends/chunkstore/chunkstore_test.go

```diff
@@ -0,0 +1,416 @@
+	}
+
+	data = make([]byte, 52)
+	test_data = test_string[:len(data)]
+	if bytes_read, err := io.ReadFull(r, data); err != nil {
+		t.Fatalf("Reading from Reader returned error for multi-chunk blob: %v", err)
+	} else if !bytes.Equal(data, test_data) {
+		t.Fatalf("Reading from Reader returned data that does not match expectation for multi-chunk blob: %v should be %v", data, test_data)
+	} else if bytes_read != len(data) {
+		t.Fatalf("Reading from Reader did not read correct number of bytes: %v should be %v", bytes_read, len(data))
+	}
+
+	overreadData = make([]byte, 52)
+	testOverreadData = append(test_string[len(data):], make([]byte, len(overreadData)+len(data)-len(test_string))...)
+	if bytes_read, err := io.ReadFull(r, overreadData); err != io.ErrUnexpectedEOF {
+		t.Fatalf("Over-reading from Reader did not return io.ErrUnexpectedEOF: %v", err)
+	} else if !bytes.Equal(overreadData, testOverreadData) {
+		t.Fatalf("Reading from Reader returned data that does not match expectation for multi-chunk blob: %v should be %v", overreadData, testOverreadData)
+	} else if bytes_read != len(test_string)-len(overreadData) {
+		t.Fatalf("Over-reading from Reader did not read enough bytes: %v should be %v", bytes_read, len(overreadData))
+	}
+
+	reverseData = make([]byte, 52)
+	testReverseData = test_string[(len(test_string) - len(reverseData)):]
+	if bytes_read, err := io.ReadFull(rr, reverseData); err != nil {
+		t.Fatalf("Reading from ReverseReader returned error for multi-chunk blob: %v", err)
+	} else if !bytes.Equal(reverseData, testReverseData) {
+		t.Fatalf("Reading from ReverseReader returned data that does not match expectation for multi-chunk blob: %v should be %v", reverseData, testReverseData)
+	} else if bytes_read != len(reverseData) {
+		t.Fatalf("Reading from ReverseReader did not read correct number of bytes: %v should be %v", bytes_read, len(reverseData))
+	}
+
+	overreadReverseData = make([]byte, 52)
+	testOverreadReverseData = append(test_string[:(len(test_string)-len(reverseData))], make([]byte, len(overreadReverseData)+len(reverseData)-len(test_string))...)
+	if bytes_read, err := io.ReadFull(rr, overreadReverseData); err != io.ErrUnexpectedEOF {
+		t.Fatalf("Over-reading from ReverseReader did not return io.ErrUnexpectedEOF: %v", err)
+	} else if !bytes.Equal(overreadReverseData, testOverreadReverseData) {
+		t.Fatalf("Reading from ReverseReader returned data that does not match expectation for multi-chunk blob: %v should be %v", overreadReverseData, testOverreadReverseData)
+	} else if bytes_read != len(test_string)-len(overreadReverseData) {
+		t.Fatalf("Over-reading from ReverseReader did not read enough bytes: %v should be %v", bytes_read, len(overreadReverseData))
+	}
+
+}
+
+func TestWriter(t *testing.T) {
+	m := mockstore.New()
+	c := New(m)
+	mtx := &mockstore.Mocktext{}
+
+	test_map := make(map[string][]byte)
+
+	w := c.Writer(mtx, "foo", 5, 50*time.Millisecond)
```

Comment:
nit: maybe define flushTime := 50*time.Millisecond, then you can do:

c.Writer(mtx, "foo", 5, flushTime)
time.Sleep(flushTime + time.Millisecond)
cmp.Equal(...)

Reply:
I admit to not having a perfect understanding of goroutines, but I'd be concerned about a test with only a millisecond buffer after another thread is supposed to have timed out might flake a lot. I have no problem with encapsulating the flush time in a variable, though.